include/LightGBM/tree.h

```diff
@@ -402,12 +427,22 @@ class Tree {
   std::vector<int> leaf_depth_;
   double shrinkage_;
   int max_depth_;
+  // add parent node information
+  std::vector<int> node_parent_;
+  // Keeps track of the monotone splits above the leaf
+  std::vector<bool> leaf_is_in_monotone_subtree_;
```

Comment:
do we really need to store these states into tree model? Or other structures only for MC? 
Since some users don't use MC, this will change the model format and cause the compatibility problem.

Reply:
another suggestion: it seems there are lots of changes in FeatureHistogram. This will hurt the readability of current codes. Could we decouple MC with other parts? Like an independent class, which could be called in FeatureHistogram.

Reply:
> This will hurt the readability of current codes. Could we decouple MC with other parts?

We were very careful to implement this in a way that limited the impact on readability and, in my opinion, I think that we did a good job in this regard, given the complexity of what's being implemented. In fact, in my opinion, the current implementation of monotonic constraints isn't particularly readable. If anything, we've made it more readable, although, obviously, more complex.

If you have any specific examples where you feel readability has been impacted let us know and we can discuss it on an example-by-example basis.

> Could we decouple MC with other parts?

We could, but I'm not convinced this would make things substantially more readable, and it would be some effort on our part to make such changes. It could even hurt readability due to increased indirection.

Reply:
Yeah, actually I mean the maintainability. All implementation of MC is in one place, not distributed in the project. And i think good maintainability is also more readable.

Reply:
I will not be able to work on this during this week and next week, sorry about that. But I will make the necessary changes to improve the code after that.

Reply:
@guolinke about the new states in the Tree class which would cause compatibility issues. First, the states are absolutely necessary to the functioning of monotone constraints as we implemented them, there is no way we can do without. And I think that these are sensible states that should belong to the tree class (I could even see these reused for other purposes). 

An alternative I can think about would be to create an alternative class which would look like tree, but with these states only, and this class would be used whenever the Tree class is currently used. But that would create some redundancies, and would probably not be good for readability. 

Since these fields currently won't be used if monotone constraints are not used, is it not possible to solve the compatibility problem instead? I am not sure when these issues would arise, but maybe there is a way to specify somewhere that if format don't match, then these fields should remain void, or be filled with dummy values. What do you think?

Reply:
@CharlesAuguste I see. could we have these states in tree class, but not save and load them into model file? If this is possible, it is okay as well.

Reply:
just double-check the code, it seems the new states are not in model file. So I think it is okay.